docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,342 @@
+template take precedence over the same parameters on the instances.
+
+The main difference with [Option 1] that enables ClusterClass is that the same ManagedCluster template
+resource can be referenced by multiple AzureManagedControlPlanes, so this new `ASOManagedClusterTemplateRef`
+field can be defined on the AzureManagedControlPlaneClassSpec so a set of AKS parameters defined once in the
+template can be applied to all Clusters built from that ClusterClass.
+
+This method makes all ManagedCluster fields available to define in a template which could lead to
+misconfiguration if certain parameters that must be unique to a cluster are erroneously shared through a
+template. Since those fields cannot be automatically identified and may evolve between AKS API versions, CAPZ
+will not attempt to categorize ASO ManagedCluster fields that way like it does between the fields present and
+omitted from the `AzureClusterClassSpec` type, for example. CAPZ could document a best-effort list of known
+fields which could or should not be defined in template types and will otherwise rely on AKS to provide
+reasonable error messages for misconfigurations.
+
+Like [Option 1], this method keeps particular versions of CAPZ decoupled from particular API versions of ASO
+resources (including allowing preview versions) and opens the door for streamlined adoption of existing AKS
+clusters.
+
+#### Option 3: CAPZ resource defines an entire unstructured ASO resource inline
+
+This method is functionally equivalent to [Option 2] except that the template resource is defined inline
+within the AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ASOManagedClusterTemplate is the ASO ManagedCluster to be used as a template from which new
+	// ManagedClusters will be created.
+	ASOManagedClusterTemplate map[string]interface{} `json:"asoManagedClusterTemplate"`
+}
+```
+
+One variant of this method could be to using `string` instead of `map[string]interface{}` for the template
+type, though that would make defining patches unwieldy (like for ClusterClass).
+
+Compared to [Option 2], this method loses schema and webhook validation that would be performed by ASO when
+creating a separate ManagedCluster to serve as a template. That validation would still be performed when CAPZ
+creates the ManagedCluster resource, but that would be some time after the AzureManagedControlPlane is created
+and error messages may not be quite as visible.
+
+#### Option 4: CAPZ resource defines an entire typed ASO resource inline
+
+This method is functionally equivalent to [Option 3] except that the template field's type is the exact same
+as an ASO ManagedCluster:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ASOManagedClusterSpec defines the spec of the ASO ManagedCluster managed by this AzureManagedControlPlane.
+	ASOManagedClusterSpec v1api20230201.ManagedCluster_Spec `json:"asoManagedClusterSpec"`
```

Comment:
minor: If I were you, I might not call this ASOManagedClusterSpec as even here in some ways ASO is an implementation detail. This is part of CAPZs API and from a future-proofing perspective rather than say `ASOManagedClusterSpec` maybe call it something like `DetailedClusterSpec` (the type can be the same).

Reason I suggest this is not because I think ASO is going away but for the same reason you probably wouldn't want to call a property `Track1SDKManagedCluster` - you're exposing an implementation detail at the API level that might not be needed.

With that said it does depend somewhat on how exactly CAPZ does API versioning. Is the interface the REST API/YAML, or is there a Go mod guarantee given as well?

Reply:
Yeah, if you're only concerned about teh JSON API, the fact this is an "ASO type" is less important and you do buy yourself quite a bit of room by dropping the ASO in the property name.